### PR TITLE
EOS-24239: change address format to support hostname fqdn for all ut's

### PR DESCRIPTION
**Pull request by kanchan-chaudhari**
Changes:
Removed unused code from libfab.c file.
Added changes to run all ut's with changed address format.

Signed-off-by: Kanchan Chaudhari <kanchan.chaudhari@seagate.com>

# Problem Statement
- **This is Work In Progress PR for Internal Review**
- change address format to support hostname fqdn for all ut's

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: net/ip.c [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+	M0_ENTRY("Hostname = %s", (char*)hostname);
+	cp = strchr(hostname, '@');
+	if (cp == NULL)
+		return M0_ERR(-EINVAL);
+
+	n = cp - hostname;
+	memcpy(name, hostname, n);
+	name[n] = '\0';
+	if ((hname = gethostbyname(name)) == NULL) {
+		/** get the host info */
```
Comment (upendrapatwardhan) [inline net/ip.c:67]: Replace '/** get the host info */'
'/* Get the host info. */'

File: net/ip.c [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+		M0_LOG(M0_ERROR, "gethostbyname error %s", (char*)name);
+		return M0_ERR(-EPROTO);
+	}
+
+	addr = (struct in_addr **) hname->h_addr_list;
+	for(i = 0; addr[i] != NULL; i++) {
+		/** Return the first one */
```
Comment (upendrapatwardhan) [inline net/ip.c:74]: Period at the end.

File: net/ip.c [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+	n = cp - hostname;
+	memcpy(name, hostname, n);
+	name[n] = '\0';
+	if ((hname = gethostbyname(name)) == NULL) {
+		/** get the host info */
+		M0_LOG(M0_ERROR, "gethostbyname error %s", (char*)name);
+		return M0_ERR(-EPROTO);
+	}
+
+	addr = (struct in_addr **) hname->h_addr_list;
```
Comment (upendrapatwardhan) [inline net/ip.c:72]: Remove space between _(struct in_addr **)_ & _hname->h_addr_list_

File: net/ip.c [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+	for(i = 0; addr[i] != NULL; i++) {
+		/** Return the first one */
+		strcpy(ip , inet_ntoa(*addr[i]));
```
Comment (upendrapatwardhan) [inline net/ip.c:75]: Space.

File: net/ip.c [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+		n=strlen(ip);
```
Comment (upendrapatwardhan) [inline net/ip.c:76]: Space.

File: net/ip.c [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+/**
+ * This function decodes the inet format address.
+ *  The inet address format is of type
+ *    <family>:<type>:<ipaddr/hostname_FQDN>@<port>
+ *    for example: "inet:tcp:127.0.0.1@3000",
+ *                 "inet:stream:lanl.gov@23",
+ *                 "inet6:dgram:FE80::0202:B3FF:FE1E:8329@6663"
+ */
+
```
Comment (upendrapatwardhan) [inline net/ip.c:95]: Remove line.

File: net/ip.c [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+	const char  *ep_name = name;
+	uint32_t     portnum;
+
+	for (f = 0; f < ARRAY_SIZE(ip_family); ++f) {
+		if (ip_family[f] != NULL) {
+			shift = strlen(ip_family[f]);
+			if (strncmp(ep_name, ip_family[f], shift) == 0)
+				break;
+		}
+	}
```
Comment (upendrapatwardhan) [inline net/ip.c:115]: Add
```
if (f < ARRAY_SIZE(ip_family))
        return M0_ERR(-EINVAL);
```

File: net/ip.c [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+	if (ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	for (s = 0; s < ARRAY_SIZE(ip_protocol); ++s) {
+		if (ip_protocol[s] != NULL) {
+			shift = strlen(ip_protocol[s]);
+			if (strncmp(ep_name, ip_protocol[s], shift) == 0)
+				break;
+		}
+	}
```
Comment (upendrapatwardhan) [inline net/ip.c:125]: Add

```
if (s < ARRAY_SIZE(ip_protocol))
        return M0_ERR(-EINVAL);
```

File: net/ip.c [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+							"stream",
+							"dgram",
+							"o2ib",
+							"lo" };
+
+/**
+ * This function convert the hostname/FQDN format to ip format.
+ * Here hostname can be FQDN or local machine hostname.
+ */
+static int m0_net_hostname_to_ip(char *hostname , char* ip,
```
Comment (upendrapatwardhan) [inline net/ip.c:48]: Remove space after hostname.

File: net/ip.c [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+			shift = strlen(ip_protocol[s]);
+			if (strncmp(ep_name, ip_protocol[s], shift) == 0)
+				break;
+		}
+	}
+	if (ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	at = strchr(ep_name, '@');
+	if (at == NULL) {
```
Comment (upendrapatwardhan) [inline net/ip.c:130]: Remove '{'.

File: net/ip.c [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+		return M0_ERR(-EINVAL);
+	} else {
+		at++;
+		if (at == NULL)
```
Comment (upendrapatwardhan) [inline net/ip.c:134]: at >= '0' && at <= '9'

File: net/ip.c [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+			return M0_ERR(-EINVAL);
+		memcpy(port, at, (strlen(at)+1));
```
Comment (upendrapatwardhan) [inline net/ip.c:136]: strcpy

File: net/ip.c [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+		portnum = atoi(port);
+		M0_ASSERT(portnum < M0_NET_IP_PORT_MAX);
+		addr->na_port = (uint16_t)portnum;
+	}
+
+	n = m0_net_hostname_to_ip((char *)ep_name, ip, addr);
+	if (n > 0)
+		memcpy(node, ip, n);
```
Comment (upendrapatwardhan) [inline net/ip.c:144]: Not required. Directly use ip instead of node in below code and remove node variable.

File: net/ip.c [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+	addr->na_addr.ia.nia_family = f;
+	addr->na_addr.ia.nia_type = s;
+
+	return 0;
+}
+
+/**
+ * Bitmap of used transfer machine identifiers. 1 is for used and 0 is for free.
+ */
+static uint8_t ip_autotm[1024] = {};
```
Comment (upendrapatwardhan) [inline net/ip.c:160]: Define along with other global variables.

File: net/ip.c [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+	}
+
+	return 0;
+}
+
+
+int m0_net_ip_parse(const char *name, struct m0_net_ip_addr *addr)
+{
+	int rc;
+	if ( name[0] >= '0' && name[0] <= '9')
```
Comment (upendrapatwardhan) [inline net/ip.c:326]: Remove extra space.

Comment (upendrapatwardhan) [reply net/ip.c:326]: M0_PRE(name != NULL);

File: net/ip.c [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+		rc = m0_net_ip_lnet_parse(name, addr);
+	else
+		rc = m0_net_ip_inet_parse(name, addr);
+	return rc;
```
Comment (upendrapatwardhan) [inline net/ip.c:330]: return (name[0] >= '0' && name[0] <= '9') ? m0_net_ip_lnet_parse(name, addr) : m0_net_ip_inet_parse(name, addr);

File: net/libfab/libfab.c [hunk 17]
```diff
@@ -357,243 +349,56 @@ M0_INTERNAL void m0_net_libfab_fini(void)
...
- *           family:type:ipaddr[@port]
- *
- *     - libfab compatible format
- *       for example IPV4 libfab:192.168.0.1:4235
- *                   IPV6 libfab:[4002:db1::1]:4235
+ *     - The inet address format is of type
+ *         <family>:<type>:<ipaddr/hostname_FQDN>@<port>
+ *        for example: "inet:tcp:127.0.0.1@3000",
+ *                     "inet:stream:lanl.gov@23",
+ *                      "inet6:dgram:FE80::0202:B3FF:FE1E:8329@6663"
```
Comment (upendrapatwardhan) [inline net/libfab/libfab.c:367]: Alignment.

File: net/libfab/libfab.c [hunk 18]
```diff
@@ -722,24 +527,42 @@ static void libfab_tm_buf_done(struct m0_fab__tm *ftm)
...
-static void libfab_straddr_gen(struct m0_fab__conn_data *cd, char *buf,
-			       uint8_t len, struct m0_fab__ep_name *en)
-{
+/** TODO: Replace this function with m0_net_ip_print() */
+static void libfab_straddr_gen(struct m0_fab__conn_data *cd,
+			       struct m0_fab__ep_name *en)
+{
+	struct m0_net_ip_addr *addr = &cd->fcd_addr;
+
+	if (addr->na_format == M0_NET_IP_LNET_FORMAT)
```
Comment (upendrapatwardhan) [inline net/libfab/libfab.c:536]: Move common code to a a function to avoid duplication.

File: net/libfab/libfab.c [hunk 19]
```diff
@@ -722,24 +527,42 @@ static void libfab_tm_buf_done(struct m0_fab__tm *ftm)
...
+		inet_ntop(AF_INET, &addr->na_n.sn[0], en->fen_addr,
+			  ARRAY_SIZE(en->fen_addr));
+	else if (addr->na_addr.ia.nia_family == M0_NET_IP_AF_INET6)
+		inet_ntop(AF_INET6, &addr->na_n.ln[0], en->fen_addr,
+			  ARRAY_SIZE(en->fen_addr));
+	else
+		M0_LOG(M0_ERROR, "Family is not supported.");
+
+	sprintf(en->fen_port, "%d", addr->na_port);
+#if 0
```
Comment (upendrapatwardhan) [inline net/libfab/libfab.c:549]: Remove unwanted code.

File: net/libfab/libfab_internal.h [hunk 20]
```diff
@@ -276,6 +276,9 @@ struct m0_fab__ep_name {
 
 	/** address in string format as passed by the net layer */
 	char fen_str_addr[LIBFAB_ADDR_STRLEN_MAX];
+
+	/** network address */
+	struct m0_net_ip_addr fen_name;
```
Comment (upendrapatwardhan) [inline net/libfab/libfab_internal.h:281]: Can we replace 'fen_str_addr' with 'fen_name.na_p'?

File: net/ip.c (hunk 2 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 1 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 3 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 4 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 5 above)
Comment (kanchan-chaudhari) [reply net/ip.c:83]: Done

File: net/ip.c (hunk 6 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 7 above)
Comment (kanchan-chaudhari) [reply net/ip.c:120]: if (f >= ARRAY_SIZE(ip_family))
        return M0_ERR(-EINVAL);
check added.

File: net/ip.c (hunk 8 above)
Comment (kanchan-chaudhari) [reply net/ip.c:132]: if (s >= ARRAY_SIZE(ip_protocol))
        return M0_ERR(-EINVAL);
check added

File: net/ip.c (hunk 9 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 10 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 11 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 12 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 13 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 14 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 15 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Added

File: net/ip.c (hunk 16 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/libfab/libfab.c (hunk 17 above)
Comment (kanchan-chaudhari) [reply net/libfab/libfab.c]: Done

File: net/libfab/libfab.c (hunk 18 above)
Comment (kanchan-chaudhari) [reply net/libfab/libfab.c]: Done

File: net/libfab/libfab.c (hunk 19 above)
Comment (kanchan-chaudhari) [reply net/libfab/libfab.c]: Done

File: net/libfab/libfab_internal.h (hunk 20 above)
Comment (kanchan-chaudhari) [reply net/libfab/libfab_internal.h]: Yes, we will take this task in further improvement, added TODO.

File: net/ip.c [hunk 21]
```diff
@@ -123,29 +130,31 @@ static int m0_net_ip_inet_parse(const char *name, struct m0_net_ip_addr *addr)
...
 		addr->na_port = (uint16_t)portnum;
 	}
 
 	n = m0_net_hostname_to_ip((char *)ep_name, ip, addr);
-	if (n > 0)
-		memcpy(node, ip, n);
-	else
-		return n == 0 ? M0_ERR(-EPROTO) : M0_ERR(n);
+	if (n == 0)
+		M0_ERR(-EPROTO);
```
Comment (upendrapatwardhan) [inline net/ip.c:153]: return M0_ERR(-EPROTO);

File: net/ip.c [hunk 22]
```diff
@@ -123,29 +130,31 @@ static int m0_net_ip_inet_parse(const char *name, struct m0_net_ip_addr *addr)
...
+	else if (n < 0)
+		M0_ERR(n);
```
Comment (upendrapatwardhan) [inline net/ip.c:155]: return M0_ERR(n);

File: net/ip.c [hunk 23]
```diff
@@ -123,29 +130,31 @@ static int m0_net_ip_inet_parse(const char *name, struct m0_net_ip_addr *addr)
...
 
-	inet_pton(AF_INET, node, &addr->na_n.sn[0]);
+	inet_pton(AF_INET, ip, &addr->na_n.sn[0]);
```
Comment (upendrapatwardhan) [inline net/ip.c:157]: Add todo here to select between IPv4 and IPv6.

File: net/ip.c (hunk 21 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 22 above)
Comment (kanchan-chaudhari) [reply net/ip.c]: Done

File: net/ip.c (hunk 23 above)
Comment (kanchan-chaudhari) [reply net/ip.c:158]: Added

**Review by upendrapatwardhan — APPROVED** [review]

**Pull request merged**